src/owners_file.rs

```diff
@@ -13,62 +14,141 @@ pub struct OwnersFileConfig {
 }
 
 impl OwnersFileConfig {
-    pub fn from_text<S0: AsRef<str>, S1: AsRef<str>>(
-        text: S0,
-        source: S1,
+    pub fn from_file<P0: AsRef<Path>, P1: AsRef<Path>>(
+        path: P0,
+        repo_base: P1,
+    ) -> anyhow::Result<OwnersFileConfig> {
+        let path_ref = path.as_ref();
+        let text = fs::read_to_string(path_ref)?;
+        Self::from_text(&text, path.as_ref(), repo_base.as_ref())
+    }
+
+    fn from_text<S: AsRef<str>, P0: AsRef<Path>, P1: AsRef<Path>>(
+        text: S,
+        path: P0,
+        repo_base: P1,
     ) -> anyhow::Result<OwnersFileConfig> {
-        let text = text.as_ref();
         let mut config = OwnersFileConfig::default();
-        let mut current_set = &mut config.all_files;
+        Self::parse_text(
+            &mut config,
+            text.as_ref(),
+            path.as_ref(),
+            repo_base.as_ref(),
+            (&mut HashMap::new()).into(),
+        )?;
+        Ok(config)
+    }
+
+    fn parse_text<P0: AsRef<Path>, P1: AsRef<Path>>(
+        config: &mut OwnersFileConfig,
+        text: &str,
+        path: P0,
+        repo_base: P1,
+        seen_owners: &mut HashMap<PathBuf, Option<PathBuf>>,
+    ) -> anyhow::Result<()> {
+        // `active_pattern_key` tracks the current context.
+        // `None`: Modifying `config.all_files`.
+        // `Some(key)`: Modifying `config.pattern_overrides` for the given key.
+        let mut active_pattern_key: Option<String> = None;
+        let source = path
+            .as_ref()
+            .to_str()
+            .expect("Error converting file path to string");
+
+        if seen_owners.is_empty() {
+            seen_owners.insert(path.as_ref().to_path_buf(), None);
+        }
 
-        for (line_number, line) in text.lines().enumerate() {
-            let line = clean_line(line);
+        for (i, raw_line) in text.lines().enumerate() {
+            let line = clean_line(raw_line);
             if line.is_empty() {
                 continue;
             }
-            let is_set_line = current_set.maybe_process_set(line).map_err(|error| {
-                anyhow!(
-                    "{} Encountered at {}:{}",
-                    error.to_string(),
-                    source.as_ref(),
-                    line_number
-                )
-            })?;
+            let line_number = i + 1;
+
+            if let Some(include_file) = maybe_get_include(line)
+                .map_err(|error| anyhow!("{} Found at {}:{}", error, source, line_number))?
+            {
+                if active_pattern_key.is_some() {
+                    return Err(anyhow!(
+                        "include is not allowed in path-specific sections. Found at {}:{}",
+                        source,
+                        line_number
+                    ));
+                }
+
+                let include_path =
+                    resolve_include_path(repo_base.as_ref(), path.as_ref(), &include_file)
```

Comment:
We don't appear to be explicitly checking for the existence of the included file. We'll get a file not found error when we try to read it, but it might be better to make the check explicit.